keps/sig-apps/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,508 @@
+The user is running a StatefulSet as part of an application with a finite lifetime. During
+the application's existence the StatefulSet maintains per-pod state, even across scale-up
+and scale-down. In order to maximize performance, volumes are retained during scale-down
+so that scale-up can leverage the existing volumes. When the application is finished, the
+volumes created by the StatefulSet are no longer needed and can be automatically
+reclaimed.
+
+The user would set the `PersistentVolumeClaimDeletePolicy` as `DeleteOnStatefulSetDelete`
+which would ensure that the PVCs created automatically during the StatefulSet activation
+is deleted once the StatefulSet is deleted.
+
+#### Story 2
+User is very cost conscious and can sustain slower scale-up speeds, even after a
+scale-down. The user does not want to pay for volumes that are not in use, and
+so wants them to be reclaimed as soon as possible, including during
+scale-down. On scale-up a new volume will be provisioned and the new pod will
+have to re-intitialize. However, for short-lived interruptions when a pod is
+killed & recreated, like a rolling update or node disruptions, the data on
+volumes is persisted. This is a key property that ephemeral storage, like
+emptyDir, cannot provide.
+
+User would set the `PersistentVolumeClaimDeletePolicy` as `DeleteOnScaledown` ensuring
+PVCs are deleted when corresponding Pods are deleted. New Pods created during scale-up
+followed by a scale-down will wait for freshly created PVCs.
+
+### Notes/Constraints/Caveats (optional)
+
+This feature applies to PVCs which are dynamically provisioned from the
+volumeClaimTemplate of a StatefulSet. Any PVC and PV provisioned from this
+mechanism will function with this feature. These PVCs are identified by the
+static naming scheme used by StatefulSets. Auto-provisioned and pre-provisioned
+PVCs will be treated identically, so that if a user pre-provisions a PVC
+matching those of a VolumeClaimTemplate it will be deleted according to the
+deletion policy.
+
+### Risks and Mitigations
+
+Currently the PVCs created by StatefulSet are not deleted automatically. Using the
+`DeleteOnScaledown` or `DeleteOnStatefulSetDeletion` would delete the PVCs
+automatically. Since this involves persistent data being deleted, users should take
+appropriate care using this feature. Having the `Retain` behaviour as default will ensure
+that the PVCs remain intact by default and only a conscious choice made by user will
+involve any persistent data being deleted.
+
+This proposed API causes the PVCs associated with the StatefulSet to have
+behavior close to, but not the same as, ephemeral volumes, such as emptyDir or
+generic ephemeral volumes. This may cause user confusion. PVCs under this policy
+will more durable than ephemeral volumes would be, as they are only deleted on
+scale-down or StatefulSet deletion, and not on other pod lifecycle events like
+being rescheduled to a new node, even with the new retain policies.
```

Comment:
I think @wojtek-t had a comment about pods not actually being rescheduled, instead they are recreated.

Reply:
Thanks, fixed this reference too.